File: backend/app/ml/pxe_filename.py
"""Edge-Impulse-style download filename builder for .pxe packages.

The canonical pattern is:

    {project-name}-{impulse-name}-v{deploy-version}.pxe

where each name component is slugified to lowercase ASCII letters, digits and
single hyphens (any other character becomes a hyphen; runs of hyphens collapse;
leading and trailing hyphens strip).
"""
from __future__ import annotations

import re

__all__ = ["slugify", "build_pxe_download_filename"]

_NON_SLUG_CHAR_RE = re.compile(r"[^a-z0-9-]+")
_MULTI_HYPHEN_RE = re.compile(r"-{2,}")


def slugify(value: str) -> str:
    """Lowercase, replace non-[a-z0-9-] runs with '-', collapse, strip."""
    if value is None:
        return ""
    s = str(value).lower()
    s = _NON_SLUG_CHAR_RE.sub("-", s)
    s = _MULTI_HYPHEN_RE.sub("-", s)
    return s.strip("-")


def build_pxe_download_filename(
    project_name: str,
    impulse_name: str,
    deploy_version: int,
) -> str:
    """Return the canonical EI-style .pxe download filename."""
    project_slug = slugify(project_name) or "project"
    impulse_slug = slugify(impulse_name) or "impulse"
    version = int(deploy_version)
    if version < 1:
        version = 1
    return f"{project_slug}-{impulse_slug}-v{version}.pxe"

File: frontend/src/lib/pxe-filename.ts
// Edge-Impulse-style download filename builder for .pxe packages.
//
// Canonical pattern:  {project-name}-{impulse-name}-v{deploy-version}.pxe
//
// Each name component is slugified to lowercase ASCII letters, digits and
// single hyphens (any other character becomes a hyphen; runs collapse; leading
// and trailing hyphens strip). Must stay in sync with backend/app/ml/pxe_filename.py.

const NON_SLUG_CHAR_RE = /[^a-z0-9-]+/g;
const MULTI_HYPHEN_RE = /-{2,}/g;

export function slugify(value: string | null | undefined): string {
  if (value == null) return "";
  return String(value)
    .toLowerCase()
    .replace(NON_SLUG_CHAR_RE, "-")
    .replace(MULTI_HYPHEN_RE, "-")
    .replace(/^-+|-+$/g, "");
}

export function buildPxeDownloadFilename(
  projectName: string,
  impulseName: string,
  deployVersion: number,
): string {
  const projectSlug = slugify(projectName) || "project";
  const impulseSlug = slugify(impulseName) || "impulse";
  const version = Math.max(1, Math.floor(Number(deployVersion) || 1));
  return `${projectSlug}-${impulseSlug}-v${version}.pxe`;
}
